src/irmin/tree.ml

```diff
@@ -852,88 +886,100 @@ module Make (P : Backend.S) = struct
-      | None -> (
-          match cached_value t with
-          | Some v -> P.Node.Val.length v |> Lwt.return
-          | None -> (
-              match t.v with
-              | Map m -> StepMap.cardinal m |> Lwt.return
-              | Key (r, k) ->
-                  value_of_key ~cache t r k
-                  >|= get_ok "length"
-                  >|= P.Node.Val.length
-              | Value (_, v, None) -> P.Node.Val.length v |> Lwt.return
-              | Value (_, v, Some u) ->
-                  hash_preimage_of_updates ~cache t v u (function
-                    | Node x -> P.Node.Val.length x |> Lwt.return
-                    | Pnode x -> P.Node_portable.length x |> Lwt.return)
-              | Pruned h -> pruned_hash_exn "length" h))
-
-    let is_empty ~cache t =
-      match cached_map t with
-      | Some m -> StepMap.is_empty m
-      | None -> (
-          match cached_value t with
-          | Some v -> P.Node.Val.is_empty v
-          | None -> (
-              match t.v with
-              | Value (_, v, Some um) -> is_empty_after_updates ~cache v um
-              | Key (_, key) -> equal_hash (P.Node.Key.to_hash key) empty_hash
-              | Pruned h -> equal_hash h empty_hash
-              | Map _ -> assert false (* [cached_map <> None] *)
-              | Value (_, _, None) -> assert false (* [cached_value <> None] *))
-          )
+    let length =
+      let dispatch t fm fv fk_repo fv_dirty fpruned =
+        iter_cached_map t fm @@ fun () ->
+        iter_cached_value t fv @@ fun () ->
+        iter_cached_repo_key t fk_repo @@ fun () ->
+        match t.v with
+        | Value (repo, v, Some um) -> fv_dirty repo v um
+        | Pruned h -> fpruned h
+        | Key _ | Map _ | Value (_, _, None) -> assert false
+      in
+      fun ~cache t ->
+        dispatch t
+          (fun m -> StepMap.cardinal m |> Lwt.return)
+          (fun v -> P.Node.Val.length v |> Lwt.return)
+          (fun repo k ->
+            value_of_key ~cache t repo k
+            >|= get_ok "length"
+            >|= P.Node.Val.length)
+          (fun _repo v um ->
+            hash_preimage_of_updates ~cache t v um (function
+              | Node x -> P.Node.Val.length x |> Lwt.return
+              | Pnode x -> P.Node_portable.length x |> Lwt.return))
+          (fun h -> pruned_hash_exn "length" h)
+
+    let is_empty =
+      let dispatch t fm fv fh fv_dirty =
+        iter_cached_map t fm @@ fun () ->
+        iter_cached_value t fv @@ fun () ->
+        iter_cached_hash t fh @@ fun () ->
+        match t.v with
+        | Value (repo, v, Some um) -> fv_dirty repo v um
+        | Map _ | Key _ | Value (_, _, None) | Pruned _ -> assert false
+      in
+      fun ~cache t ->
+        dispatch t
+          (fun m -> StepMap.is_empty m)
+          (fun v -> P.Node.Val.is_empty v)
+          (fun h -> equal_hash h empty_hash)
+          (fun _repo v um -> is_empty_after_updates ~cache v um)
```

Comment:
Did you consider labelling the continuation arguments like this?

```suggestion
      let dispatch t ~map ~value ~hash ~dirty =
        iter_cached_map t fm @@ fun () ->
        iter_cached_value t fv @@ fun () ->
        iter_cached_hash t fh @@ fun () ->
        match t.v with
        | Value (repo, v, Some um) -> fv_dirty repo v um
        | Map _ | Key _ | Value (_, _, None) | Pruned _ -> assert false
      in
      fun ~cache t ->
        dispatch t
          ~map:(fun m -> StepMap.is_empty m)
          ~value:(fun v -> P.Node.Val.is_empty v)
          ~hash:(fun h -> equal_hash h empty_hash)
          ~dirty:(fun _repo v um -> is_empty_after_updates ~cache v um)
```

I don't have a strong preference; just wondering.